src/webgpu/web_platform/canvas/getCurrentTexture.spec.ts

```diff
@@ -260,3 +277,103 @@ g.test('resize')
+  - in a microtask off the current frame task
+- getCurrentTexture returns a new texture object and the old texture object becomes invalid
+  as soon as possible after HTML update the rendering.
+
+TODO: make [1] a warning instead of error.
+
+TODO: test more canvas types, and ways to update the rendering
+- if on the same thread, expiry happens when the document updates its rendering (window "rPAF") OR transferToImageBitmap is called
+- if on a different thread, expiry happens when the worker updates its rendering (worker "rPAF") OR transferToImageBitmap is called
+- [draw, transferControlToOffscreen, then canvas is displayed] on either {main thread, or transferred to worker}
+- [draw, canvas is displayed, then transferControlToOffscreen] on either {main thread, or transferred to worker}
+- reftests for the above 2 (what gets displayed when the canvas is displayed)
+- with canvas element added to DOM or not (applies to other canvas tests as well)
+  - canvas is added to DOM after being rendered
+  - canvas is already in DOM but becomes visible after being rendered
+  `
+  )
+  .params(u =>
+    u //
+      .combine('canvasType', kAllCanvasTypes)
+      .combine('prevFrameCallsite', ['runInNewCanvasFrame', 'requestAnimationFrame'] as const)
+      .combine('getCurrentTextureAgain', [true, false] as const)
+  )
+  .fn(t => {
+    const { canvasType, prevFrameCallsite, getCurrentTextureAgain } = t.params;
+    const ctx = t.initCanvasContext(t.params.canvasType);
+    // Create a bindGroupLayout to test invalid texture view usage later.
+    const bgl = t.device.createBindGroupLayout({
+      entries: [
+        {
+          binding: 0,
+          visibility: GPUShaderStage.COMPUTE,
+          texture: {},
+        },
+      ],
+    });
+
+    // The fn is called immediately after previous frame updating the rendering.
+    // Polyfill by calling the callback by setTimeout, in the requestAnimationFrame callback (for onscreen canvas)
+    // or after transferToImageBitmap (for offscreen canvas).
+    function runInNewCanvasFrame(fn: () => void) {
+      t.requestNewFrameOnCanvasType(canvasType, ctx, () => {
+        if (canvasType === 'onscreen') {
+          timeout(fn);
+        } else {
+          // for offscreen canvas, after calling transferToImageBitmap, we are in a new frame immediately
+          fn();
+        }
+      });
+    }
+
+    function checkGetCurrentTexture() {
+      // Call getCurrentTexture on previous frame.
+      const prevTexture = ctx.getCurrentTexture();
+
+      // Call getCurrentTexture immediately after the frame, the texture object should stay the same.
+      queueMicrotask(() => {
+        if (getCurrentTextureAgain) {
+          t.expect(prevTexture === ctx.getCurrentTexture());
+        }
+      });
+
+      // Call getCurrentTexture immediately after this frame updating the rendering.
+      // It could return a new texture object as early as the next task. [1]
+      timeout(() => {
```

Comment:
In the `prevFrameCallsite="runInNewCanvasFrame"` case, I think this timeout() happens during the same frame as the first getCurrentTexture() (line 341). runInNewCanvasFrame() instead of timeout() here should work?

Reply:
wait, I misunderstood something. Let me get back to you about this

Reply:
OK I understand, this problem is the TODO about making it a warning. Let's not worry about the warning anymore since we will write a chrome-specific test, and just write the spec-compliant test.

So replacing this timeout() with runInNewCanvasFrame() seems right.

Reply:
Turns out changing this `timeout` to `runInNewCanvasFrame`, makes that when canvas is offscreen, the new frame fn is called before the `queueMicroTask` fn, because offscreen canvas new frame method `transferToImageBitmap` returns immediately.

I can think of putting this `runInNewCanvasFrame` inside the fn of `queueMicroTask` to pass the test. Do you think this still keeps the semantic we want?

Reply:
That's probably fine. Theoretically we would also want to test transferToImageBitmap without any task boundaries in between but that could be another test, or not at all. We might need another test anyway if we want to test things like transferToImageBitmap interaction with presentation of a transferControlToOffscreen'd canvas.